### PR TITLE
use more cython features for item.pyx

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
this is a simple transform to c level classes, which removes the need for slots, and uses c level properties as well

i wonder how to benchmark the impact

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
Thanks. Did you see some tests are failing?

What's the supposed performance (speed, memory usage) impact of this?

**Comment by RonnyPfannschmidt** [comment]
i wonder how to measure, and im currently fighting the memory views for the chunk compare

**Comment by ThomasWaldmann** [comment]
Speed: There are some tests using pytest-benchmark, there is `borg bench(mark?)` command or one could just run borg against a specific dataset on a idle machine, backing up into a local repo.

Memory usage: not sure we have something for that yet.

**Comment by RonnyPfannschmidt** [comment]
i finished a initial iteraation, im going to work out in more detail some properties

i want method tables to be gone
i want the encoded /optional properties to use c level encode/decode

File: src/borg/item.pyx [hunk 1]
```diff
@@ -528,35 +530,68 @@ class ItemDiff:
...
+        return  _chunk_content_equal(chunks1, chunks2)
     
+
+cdef inline size_t min_slice(size_t a, size_t b):
+    return a if a < b else b
+
+@cython.final
+cdef  class chunk_compare_state:
+    # TODO: remove the method table
+      cdef Py_ssize_t len
```
Comment (ThomasWaldmann) [inline src/borg/item.pyx:545]: indentation looks unequal here. tabs?

File: src/borg/item.pyx [hunk 2]
```diff
@@ -528,35 +530,68 @@ class ItemDiff:
...
+    @staticmethod
+    def _chunk_content_equal(chunks1, chunks2):
+        return  _chunk_content_equal(chunks1, chunks2)
     
+
+cdef inline size_t min_slice(size_t a, size_t b):
+    return a if a < b else b
+
+@cython.final
+cdef  class chunk_compare_state:
```
Comment (ThomasWaldmann) [inline src/borg/item.pyx:543]: sometimes you have 2 spaces after cdef, sometimes 1.

File: src/borg/item.pyx [hunk 3]
```diff
@@ -528,35 +530,68 @@ class ItemDiff:
...
     Compare chunk content and return True if they are identical.
 
     The chunks must be given as chunk iterators (like returned by :meth:`.DownloadPipeline.fetch_many`).
     """
 
-    end = object()
-    cdef size_t alen = 0, ai = 0
-    cdef size_t blen = 0, bi = 0
-
+    cdef chunk_compare_state s1=chunk_compare_state(chunks1), s2= chunk_compare_state(chunks2)
```
Comment (ThomasWaldmann) [inline src/borg/item.pyx:586]: ?

File: src/borg/item.pyx [hunk 4]
```diff
@@ -528,35 +530,68 @@ class ItemDiff:
...
-        ai += slicelen
-        bi += slicelen
+        if not s1.slicelen():
+            return s2.is_consumed_destructive()
+        if not s2.slicelen():
+            return s1.is_consumed_destructive()
+            
+        slicelen = min_slice(s1.slicelen(), s2.slicelen())
+        if memcmp(s1.current(), s2.current(), slicelen) != 0:
+          return False
```
Comment (ThomasWaldmann) [inline src/borg/item.pyx:595]: indentation?

**Review by ThomasWaldmann — COMMENTED** [review]
some nitpicks / strange looking stuff i found.

File: src/borg/item.pyx (hunk 3 above)
Comment (RonnyPfannschmidt) [reply src/borg/item.pyx:586]: hose abstract the iteration state, that helps to make sense of the loop details

**Comment by RonnyPfannschmidt** [comment]
i havent figured why src/borg/testsuite/archiver.py::DiffArchiverTestCase::test_basic_functionality fails yet

File: src/borg/item.pyx [hunk 5]
```diff
@@ -393,16 +406,27 @@ class ManifestItem(PropDict):
...
-    config = PropDict._make_property('config', dict)
-    item_keys = PropDict._make_property('item_keys', tuple)
+cpdef _init_names():
+    cdef PropDictProperty val = None
+    for cls in PropDict.__subclasses__():
+       for name, value  in vars(cls).items():
+          if isinstance(value, PropDictProperty):
+            val = value
+            print(cls, name, val)
+
```
Comment (ThomasWaldmann) [inline src/borg/item.pyx:426]: remove?

File: src/borg/testsuite/archiver.py [hunk 6]
```diff
@@ -3996,6 +3996,7 @@ def test_basic_functionality(self):
 
         def do_asserts(output, can_compare_ids):
             # File contents changed (deleted and replaced with a new file)
+            print(output)
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py:3999]: remove?

**Review by ThomasWaldmann — COMMENTED** [review]
some nitpicks. there are also some pep8/spacing issues like double-spaces or double-newlines.

File: src/borg/item.pyx [hunk 7]
```diff
@@ -94,44 +96,57 @@ class PropDict:
...
-
-        def _del(self):
-            try:
-                del self._dict[key]
-            except KeyError:
-                raise AttributeError(attr_error_msg) from None
 
-        return property(_get, _set, _del, doc=doc)
+cdef class PropDictProperty:
+    "return a property that deals with self._dict[key] of  PropDict"
```
Comment (ThomasWaldmann) [inline src/borg/item.pyx:101]: """..."""

Comment (ThomasWaldmann) [reply src/borg/item.pyx:101]: I know it also works with single, but triple-double-quotes is the usual style we use in the code.

File: src/borg/item.pyx [hunk 8]
```diff
@@ -517,36 +535,37 @@ class ItemDiff:
...
+        return _chunk_content_equal(chunk_iterator1, chunk_iterator2)
 
-    @staticmethod
-    def _chunk_content_equal(chunks1, chunks2):
-        """
-        Compare chunk content and return True if they are identical.
 
-        The chunks must be given as chunk iterators (like returned by :meth:`.DownloadPipeline.fetch_many`).
-        """
+def _chunk_content_equal(chunks1, chunks2):
```
Comment (ThomasWaldmann) [inline src/borg/item.pyx:541]: if that's imported from elsewhere, guess we should remove the leading underscore?

File: src/borg/item.pyx [hunk 9]
```diff
@@ -517,36 +535,37 @@ class ItemDiff:
...
+            if b is None:
+                return not alen and next(chunks1, None) is None
+            PyBytes_AsStringAndSize(b, &bp, &blen)
+        slicelen = min(alen, blen)
+        if memcmp(ap, bp, slicelen) != 0:
+            return False
+        ap += slicelen
+        bp += slicelen
+        alen -= slicelen
+        blen -= slicelen
```
Comment (ThomasWaldmann) [inline src/borg/item.pyx:571]: looks good!

one minor nitpick is naming consistency (was already there before your PR, but maybe we could fix it right now, too):
```
chunks1 -> achunks
chunks2 -> bchunks
```
Guess this is the minimal change to get it consistent.

File: src/borg/item.pyx (hunk 8 above)
Comment (ThomasWaldmann) [reply src/borg/item.pyx:541]: also, as it is not just comparing one chunk, but multiple, should it be `chunks_contents_equal`?

**Comment by RonnyPfannschmidt** [comment]
ok, unless i missed something thats done

now whats missing in top is a script to measure perf changes, i didnt find the time to look into that

File: src/borg/item.pyx [hunk 10]
```diff
@@ -393,16 +400,27 @@ class ManifestItem(PropDict):
...
-    item_keys = PropDict._make_property('item_keys', tuple)
+cpdef _init_names():
+    """
+    re-implements python __set_name__
+    """
+    cdef PropDictProperty val = None
+    for cls in PropDict.__subclasses__():
+         for name, value  in vars(cls).items():
+             if isinstance(value, PropDictProperty):
+                 val = value
```
Comment (ThomasWaldmann) [inline src/borg/item.pyx:420]: what's the point of this?

Comment (RonnyPfannschmidt) [reply src/borg/item.pyx:420]: Access to c levels attributes, but I'm thinking about having readonly public there

**Comment by ThomasWaldmann** [comment]
@RonnyPfannschmidt can i fix some cosmetic issues?

**Comment by RonnyPfannschmidt** [comment]
@ThomasWaldmann feel free to, may i funnel up the rename of the chunk content compare function to a global to master, it makes building the benchmarks against master vs this easier, and i believe for more practical speed-up i might need to add some extras for the encoding down the line (aka turning the python lookups into a c vtable lookup

for a few personal/professional reasons i may not be able to get back to this until after the python language summit

**Comment by ThomasWaldmann** [comment]
ok, pep8 issues fixed, so the linter does not block the real tests any more.

**Comment by ThomasWaldmann** [comment]
I'ld like to make a release soon. Would be nice to have this merged, just make sure it is no regression with resource consumption.

**Comment by RonnyPfannschmidt** [comment]
@ThomasWaldmann i cant yet quantify if this add sensible/measurable perf gains, but im happy to rebase later today to make it mergable and recent

**Comment by RonnyPfannschmidt** [comment]
@ThomasWaldmann i finally got around rebasing this, sorry for the delay, i think one of the next steps is me taking a good look at profiles

**Comment by ThomasWaldmann** [comment]
Any progress here? As this is a low-level change, I'ld rather like to have this merged before releasing a 1.2 release-candidate.

**Comment by ThomasWaldmann** [comment]
Also, tests are failing in a strange way - maybe rebase on current master and force-push to trigger a new test run?

**Comment by RonnyPfannschmidt** [comment]
thanks for the ping, will rebase this evening

**Comment by ThomasWaldmann** [comment]
BTW, I have only 1 ticket  (#5679) left for next milestone (which is 1.2.0rc1).

After rc1, I would not like to add any low-level changes, so this PR would have to wait until 1.2.1(beta/rc) at least.

The minimum required for merge is working unit test and **some** reasoning why this is at least as good as what we have from the performance / resource usage perspective (== no regression).

I'ld like to have it in as the source code is prettier.

**Comment by RonnyPfannschmidt** [comment]
For personal reasons I won't be able to touch this with the next 4 weeks at least 🙁

**Comment by ThomasWaldmann** [comment]
Well, my timing also came out differently than expected. But the current idea is to get 1.2 out at twos-days (2022-02-22).

So if this can be updated / fixed, we could try to get it into final beta or rc1.

**Comment by ThomasWaldmann** [comment]
I'ld try a rebase on current master now (if you don't mind) to see if the test failures still happen after that.

**Comment by ThomasWaldmann** [comment]
Rebased without any conflicts, so let's see what CI says...

**Comment by ThomasWaldmann** [comment]
CI looking better now.

**Comment by RonnyPfannschmidt** [comment]
thanks for the rebase/fix, im still far from taking a look myself, i hope to make the escaping helpers cimport-able sooner or later to sidestep python there as well
but those bits can be skipped for getting the normal enhancement in first

**Comment by ThomasWaldmann** [comment]
Please do not add more features here.

Blocking here is the evaluation of performance / resource usage of the existing proposed changes (== assuring there is no regression).

**Comment by ThomasWaldmann** [comment]
see #7063.

**Pull request closed without merging**